File: 143-reorder-list/reorder-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if(!head || head->next==NULL) return head;
        ListNode* nhead=reverseList(head->next);
        head->next->next=head;
        head->next=NULL;
        return nhead;
    }
    void reorderList(ListNode* head) {
        if(!head || head->next==NULL || head->next->next==NULL) return ;
        int len=0;
        ListNode* lft=head;
        while(lft){
            ++len;
            lft=lft->next;
        }
        ListNode *rgt=NULL, *temp=head;
        for(int i=0;i<(len+1)/2-1;++i){
            temp=temp->next;
        }
        rgt=temp->next;
        temp->next=NULL;
        lft=head;
        rgt=reverseList(rgt);
        ListNode *nhead=NULL;
        temp=NULL;
        while(lft){
            if(!nhead){
                nhead=lft;
                temp=lft;
                lft=lft->next;
                
            }
            else{
                temp->next=lft;
                temp=temp->next;
                lft=lft->next;
            }
            if(rgt){
                temp->next=rgt;
                temp=temp->next;
                rgt=rgt->next;
            }
            temp->next=NULL;
        }
        head= nhead;
    }
};

File: 143-reorder-list/reorder-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    private ListNode reverseLL(ListNode head){
        ListNode curr = head, nxt= null, prev= null;
        while(curr!=null){
            nxt= curr.next;
            curr.next= prev;
            prev= curr;
            curr= nxt;
        }
        return prev;
    }
    private void printLL(ListNode list){
        while(list!=null){
            System.out.print(list.val+" ");
            list= list.next;
        }
        System.out.println();
    }
    private ListNode merge(ListNode a, ListNode b){
        if(a==null && b==null) return null;
        else if(b==null) return a;
        ListNode head= a;
        ListNode curr= a; a= a.next;
        curr.next= b; curr= curr.next;
        b= b.next; curr.next= null;
        curr.next= merge(a, b);
        return head;
    }
    public void reorderList(ListNode head) {
        if(head== null || head.next== null || head.next.next==null) return;
        ListNode slow= head, fast= head;
        while(fast.next!=null && fast.next.next!=null){
            slow= slow.next;
            fast= fast.next.next;
        }
        ListNode list1= head;
        ListNode list2= slow.next;
        slow.next= null;
        list2= reverseLL(list2);
        // printLL(list1); printLL(list2);
        head= merge(list1, list2);
    }
}